File: ctci/tmp/chapter_4_trees_graphs/js/tree.js
class TreeNode {
  /**
   * Tree Node constructor.
   * @param   {number}  value
   */
  constructor(value) {
    this.value = value
    this.left = null
    this.right = null
    this.parent = null
  }

  /**
   * Adds a node to the left.
   * @param   {number} value
   */
  addLeft(value) {
    const new_node = new TreeNode(value)
    new_node.parent = this
    this.left = new_node
    return new_node
  }
  /**
   * Adds a node to the right.
   * @param   {number} value
   */
  addRight(value) {
    const new_node = new TreeNode(value)
    new_node.parent = this
    this.right = new_node
    return new_node
  }
}

/**
 * Prints current node before child nodes.
 * @param   {TreeNode} node
 */
const preOrderTraversal = node => {
  if (!node) return
  console.log(node.value)
  preOrderTraversal(node.left)
  preOrderTraversal(node.right)
}

/**
 * Prints left nodes, current node, and then right nodes.
 * @param   {TreeNode} node
 */
const inOrderTraversal = node => {
  if (!node) return
  inOrderTraversal(node.left)
  console.log(node.value)
  inOrderTraversal(node.right)
}

/**
 * Visits child nodes before current node.
 * @param   {TreeNode} node
 */
const postOrderTraversal = node => {
  if (!node) return
  postOrderTraversal(node.left)
  postOrderTraversal(node.right)
  console.log(node.value)
}

module.exports = {
  TreeNode,
  preOrderTraversal,
  inOrderTraversal,
  postOrderTraversal
}

File: ctci/tmp/chapter_4_trees_graphs/python/tree.py
class TreeNode:
    def __init__(self, value: int):
        """
        Tree Node constructor.
        """
        self.value = value
        self.parent = None
        self.left = None
        self.right = None

    def add_left(self, value: int):
        """
        Adds a node to the left.
        """
        new_node = TreeNode(value)
        new_node.parent = self
        self.left = new_node
        return new_node

    def add_right(self, value: int):
        """
        Adds a node to the right.
        """
        new_node = TreeNode(value)
        new_node.parent = self
        self.right = new_node
        return new_node


def pre_order_traversal(node: TreeNode):
    """
    Prints current node before child nodes.
    """
    if not node:
        return
    print(node.value)
    pre_order_traversal(node.left)
    pre_order_traversal(node.right)


def in_order_traversal(node: TreeNode):
    """
    Prints left nodes, current node, and then right nodes.
    """
    if not node:
        return
    in_order_traversal(node.left)
    print(node.value)
    in_order_traversal(node.right)


def post_order_traversal(node: TreeNode):
    """
    Visits child nodes before current node.
    """
    if not node:
        return
    post_order_traversal(node.left)
    post_order_traversal(node.right)
    print(node.value)
